File: leetcode/623-add-one-row-to-tree.js
// https://leetcode.com/problems/add-one-row-to-tree/

// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @param {number} val
 * @param {number} depth
 * @return {TreeNode}
 */
var addOneRow = function(root, val, depth) {
  if (depth === 1) return new TreeNode(val, root);

  const dfs = (node, currDepth = 1) => {
    if(!node) return;
    if (currDepth === depth - 1) {
      const leftTreeNode = new TreeNode(val, node.left);
      const rightTreeNode = new TreeNode(val, null, node.right);
      node.left = leftTreeNode;
      node.right = rightTreeNode;
      return;
    }

    dfs(node.left, currDepth + 1);
    dfs(node.right, currDepth + 1);
  }

  dfs(root);
  return root;
};

File: leetcode/623-add-one-row-to-tree.py
# https://leetcode.com/problems/add-one-row-to-tree/

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1: return TreeNode(val, root)

        def dfs(node, curr_depth = 1):
            if not node: return
            if curr_depth == depth - 1:
                node.left = TreeNode(val, node.left)
                node.right = TreeNode(val, None, node.right)
                return node
            dfs(node.left, curr_depth + 1)
            dfs(node.right, curr_depth + 1)
            return node
            
        return dfs(root)
